File: src/backend/src/main/java/com/mimimart/application/service/AddressService.java
package com.mimimart.application.service;

import com.mimimart.domain.member.exception.AddressNotFoundException;
import com.mimimart.infrastructure.persistence.entity.MemberAddress;
import com.mimimart.infrastructure.persistence.repository.MemberAddressRepository;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

/**
 * 收貨地址服務
 *
 * @author MimiMart Development Team
 * @since 1.0.0
 */
@Service
public class AddressService {

    private final MemberAddressRepository addressRepository;

    public AddressService(MemberAddressRepository addressRepository) {
        this.addressRepository = addressRepository;
    }

    /**
     * 查詢會員的所有地址
     */
    public List<MemberAddress> getAddressList(Long memberId) {
        return addressRepository.findByMemberId(memberId);
    }

    /**
     * 新增地址
     */
    @Transactional
    public MemberAddress createAddress(Long memberId, String recipientName, String phone, String address, Boolean isDefault) {
        // 如果設為預設,先取消其他預設地址
        if (Boolean.TRUE.equals(isDefault)) {
            addressRepository.findByMemberIdAndIsDefaultTrue(memberId)
                    .ifPresent(defaultAddr -> {
                        defaultAddr.setIsDefault(false);
                        addressRepository.save(defaultAddr);
                    });
        }

        MemberAddress memberAddress = new MemberAddress();
        memberAddress.setMemberId(memberId);
        memberAddress.setRecipientName(recipientName);
        memberAddress.setPhone(phone);
        memberAddress.setAddress(address);
        memberAddress.setIsDefault(isDefault != null ? isDefault : false);

        return addressRepository.save(memberAddress);
    }

    /**
     * 更新地址
     */
    @Transactional
    public MemberAddress updateAddress(Long memberId, Long addressId, String recipientName, String phone, String address) {
        MemberAddress memberAddress = addressRepository.findByIdAndMemberId(addressId, memberId)
                .orElseThrow(() -> new AddressNotFoundException("地址不存在"));

        if (recipientName != null) {
            memberAddress.setRecipientName(recipientName);
        }
        if (phone != null) {
            memberAddress.setPhone(phone);
        }
        if (address != null) {
            memberAddress.setAddress(address);
        }

        return addressRepository.save(memberAddress);
    }

    /**
     * 刪除地址
     */
    @Transactional
    public void deleteAddress(Long memberId, Long addressId) {
        MemberAddress memberAddress = addressRepository.findByIdAndMemberId(addressId, memberId)
                .orElseThrow(() -> new AddressNotFoundException("地址不存在"));

        addressRepository.delete(memberAddress);
    }

    /**
     * 設為預設地址
     */
    @Transactional
    public void setDefaultAddress(Long memberId, Long addressId) {
        // 取消其他預設地址
        addressRepository.findByMemberIdAndIsDefaultTrue(memberId)
                .ifPresent(defaultAddr -> {
                    defaultAddr.setIsDefault(false);
                    addressRepository.save(defaultAddr);
                });

        // 設定新的預設地址
        MemberAddress memberAddress = addressRepository.findByIdAndMemberId(addressId, memberId)
                .orElseThrow(() -> new AddressNotFoundException("地址不存在"));

        memberAddress.setIsDefault(true);
        addressRepository.save(memberAddress);
    }
}

File: src/shop-frontend/src/services/addressService.js
/**
 * 收貨地址相關 API 服務
 */
import apiClient from './api';

export const addressService = {
  /**
   * 查詢收貨地址列表
   * @returns {Promise} 地址列表
   */
  async getAddressList() {
    return await apiClient.get('/api/shop/address/list');
  },

  /**
   * 新增收貨地址
   * @param {Object} data - 地址資料
   * @param {string} data.recipientName - 收件人姓名
   * @param {string} data.phone - 收件人電話
   * @param {string} data.address - 收貨地址
   * @param {boolean} data.isDefault - 是否為預設地址
   * @returns {Promise} 新增結果
   */
  async createAddress(data) {
    return await apiClient.post('/api/shop/address/create', data);
  },

  /**
   * 更新收貨地址
   * @param {Object} data - 地址資料
   * @param {number} data.addressId - 地址 ID
   * @param {string} data.recipientName - 收件人姓名
   * @param {string} data.phone - 收件人電話
   * @param {string} data.address - 收貨地址
   * @param {boolean} data.isDefault - 是否為預設地址
   * @returns {Promise} 更新結果
   */
  async updateAddress(data) {
    return await apiClient.post('/api/shop/address/update', data);
  },

  /**
   * 刪除收貨地址
   * @param {number} addressId - 地址 ID
   * @returns {Promise} 刪除結果
   */
  async deleteAddress(addressId) {
    return await apiClient.post('/api/shop/address/delete', { addressId });
  },

  /**
   * 設為預設地址
   * @param {number} addressId - 地址 ID
   * @returns {Promise} 設定結果
   */
  async setDefaultAddress(addressId) {
    return await apiClient.post('/api/shop/address/set-default', { addressId });
  },
};
